azure-iot-device/tests/common/pipeline/test_pipeline_stages_base.py

```diff
@@ -107,12 +99,441 @@ def mock_execute_op(self, op):
+
+        assert stage.next.run_op.call_count == 2
+        assert stage.next.run_op.call_args[0][0] == fake_op
+
+    @pytest.mark.parametrize(
+        "params", connect_ops, ids=[x["op_class"].__name__ for x in connect_ops]
+    )
+    @pytest.mark.it("Fails the operation if currently serialized option fails")
+    def test_fails_blocked_op_if_serialized_op_fails(
+        self, params, stage, connect_op, fake_op, fake_exception
+    ):
+        stage.pipeline_root.connected = params["connected_flag_required_to_run"]
+        stage.run_op(connect_op)
+        stage.run_op(fake_op)
+        connect_op.error = fake_exception
+        operation_flow.complete_op(stage=stage.next, op=connect_op)
+        assert_callback_failed(op=fake_op, error=fake_exception)
+
+    @pytest.mark.parametrize(
+        "params", connect_ops, ids=[x["op_class"].__name__ for x in connect_ops]
+    )
+    @pytest.mark.it("Can pend multiple operations while waiting for a serialized operation")
+    def test_blocks_multiple_ops(self, params, stage, connect_op, fake_ops):
+        stage.pipeline_root.connected = params["connected_flag_required_to_run"]
+        stage.run_op(connect_op)
+        for op in fake_ops:
+            stage.run_op(op)
+        assert stage.next.run_op.call_count == 1
+
+    @pytest.mark.parametrize(
+        "params", connect_ops, ids=[x["op_class"].__name__ for x in connect_ops]
+    )
+    @pytest.mark.it("Passes down all pending operations after the serialized operation completes")
+    def test_unblocks_multiple_ops(self, params, stage, connect_op, fake_ops):
+        stage.pipeline_root.connected = params["connected_flag_required_to_run"]
+        stage.run_op(connect_op)
+        for op in fake_ops:
+            stage.run_op(op)
+
+        operation_flow.complete_op(stage=stage.next, op=connect_op)
+
+        assert stage.next.run_op.call_count == 1 + len(fake_ops)
+
+        # zip our ops and our calls together and make sure they match
+        run_ops = zip(fake_ops, stage.next.run_op.call_args_list[1:])
+        for run_op in run_ops:
+            op = run_op[0]
+            call_args = run_op[1]
+            assert op == call_args[0][0]
+
+    @pytest.mark.parametrize(
+        "params", connect_ops, ids=[x["op_class"].__name__ for x in connect_ops]
+    )
+    @pytest.mark.it("Fails all pending operations after the seriazed operation fails")
+    def test_fails_multiple_ops(self, params, stage, connect_op, fake_ops, fake_exception):
+        stage.pipeline_root.connected = params["connected_flag_required_to_run"]
+        stage.run_op(connect_op)
+        for op in fake_ops:
+            stage.run_op(op)
+
+        connect_op.error = fake_exception
+        operation_flow.complete_op(stage=stage.next, op=connect_op)
+
+        for op in fake_ops:
+            assert_callback_failed(op=op, error=fake_exception)
+
+    @pytest.mark.it(
+        "Does not pass down operations in the queue if an operation in the queue gets serialized"
```

Comment:
>Does not pass down operations in the queue if an operation in the queue gets serialized [](start = 9, length = 87)

This is a hard one to phrase!

Does not pass down further operations in the queue once an operation in the queue begins serialization?